libs/framework/src/module.c

```diff
@@ -319,4 +351,172 @@ array_list_pt module_getDependents(module_pt module) {
+
+        while (pathToken != NULL) {
+
+            /*Disable version should be part of the lib name
+            if (strncmp(pathToken, "version", 7) == 0) {
+                char *ver = strdup(pathToken);
+                char version[strlen(ver) - 9];
+                strncpy(version, ver+9, strlen(ver) - 10);
+                version[strlen(ver) - 10] = '\0';
+
+                strcat(lib, "-");
+                strcat(lib, version);
+            }*/
+            pathToken = strtok_r(NULL, ";", &path);
+        }
+
+        char *trimmedLib = utils_stringTrim(lib);
+        status = celix_module_loadLibraryForManifestEntry(module, trimmedLib, archive, &handle);
+
+        if ( (status == CELIX_SUCCESS) && (activator != NULL) && (strcmp(trimmedLib, activator) == 0) ) {
+            *activatorHandle = handle;
+        }
+        else if (handle!=NULL) {
+            celix_libloader_close(fwCtx, handle);
+        }
+
+        token = strtok_r(NULL, ",", &last);
+    }
+
+    free(libraries);
+    return status;
+}
+
+celix_status_t celix_module_loadLibraries(celix_module_t* module) {
+    celix_status_t status = CELIX_SUCCESS;
+    celix_bundle_context_t* fwCtx = celix_framework_getFrameworkContext(module->fw);
+
+    celix_library_handle_t* activatorHandle = NULL;
+    bundle_archive_pt archive = NULL;
+    bundle_revision_pt revision = NULL;
+    manifest_pt manifest = NULL;
+
+    status = CELIX_DO_IF(status, bundle_getArchive(module->bundle, &archive));
+    status = CELIX_DO_IF(status, bundleArchive_getCurrentRevision(archive, &revision));
+    status = CELIX_DO_IF(status, bundleRevision_getManifest(revision, &manifest));
+    if (status == CELIX_SUCCESS) {
+        const char *privateLibraries = NULL;
+        const char *exportLibraries = NULL;
+        const char *activator = NULL;
+
+        privateLibraries = manifest_getValue(manifest, CELIX_FRAMEWORK_PRIVATE_LIBRARY);
+        exportLibraries = manifest_getValue(manifest, CELIX_FRAMEWORK_EXPORT_LIBRARY);
+        //@note not import yet
+        activator = manifest_getValue(manifest, CELIX_FRAMEWORK_BUNDLE_ACTIVATOR);
+
+        if (exportLibraries != NULL) {
+            status = CELIX_DO_IF(status, celix_module_loadLibrariesInManifestEntry(module, exportLibraries, activator, archive, &activatorHandle));
+        }
+
+        if (privateLibraries != NULL) {
+            status = CELIX_DO_IF(status,
+                                 celix_module_loadLibrariesInManifestEntry(module, privateLibraries, activator, archive, &activatorHandle));
+        }
+
+        if (status == CELIX_SUCCESS) {
+            bundle_setHandle(module->bundle, activatorHandle); //note deprecated
+            celixThreadMutex_lock(&module->handlesLock);
+            module->bundleActivatorHandle = activatorHandle;
+            celixThreadMutex_unlock(&module->handlesLock);
+        } else if (activatorHandle != NULL) {
+            celix_libloader_close(fwCtx, activatorHandle);
```

Comment:
Double-close?

Reply:
No, because the else if only happens if status != CELIX_SUCCES.

Reply:
I still feel bad about the current code.

Every handle  come from a successful `celix_libloader_open`, which will add the handle into `module->libraryHandles, libHandle`. When `celix_module_closeLibraries` is called, these handles will be closed. Any `celix_libloader_close` without removing a handle from `module->libraryHandles, libHandle` should be viewed as double-close. 

In other words, there should be no `celix_libloader_close` in this file.

Reply:
> No, because the else if only happens if status != CELIX_SUCCES.

If this branch is not dead code (which should be removed completely without any harm), there must be a double-close.